File: solution/103BinaryTreeZigzagLevelOrderTraversal.cpp
//
// Created by Alaric on 2019-10-10.
//
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if (root == NULL)
            return ans;
        bool flag = true;
        queue<TreeNode*> q;
        q.push(root);
        while (!q.empty()){
            int n = q.size();
            vector<int> cur_level;
            for(int i = 0 ; i < n ; ++i){
                TreeNode* cur_node = q.front();
                q.pop();
                if (cur_node -> left != NULL)
                    q.push(cur_node->left);
                if (cur_node -> right != NULL)
                    q.push(cur_node->right);
                cur_level.push_back(cur_node->val);
            }
            if(!flag)
                reverse(cur_level.begin() , cur_level.end());
            flag = !flag;
            ans.push_back(cur_level);
        }
        return ans;
    }
};

File: solution/103BinaryTreeZigzagLevelOrderTraversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        queue = [root]
        ans = []
        flag = False
        while queue:
            flag = bool(1-flag)
            cur_level = []
            n = len(queue)
            for i in range(n):
                temp = queue.pop(0)
                if temp.left : queue.append(temp.left)
                if temp.right : queue.append(temp.right)
                cur_level.append(temp.val)
            if flag :
                ans.append(cur_level)
            else:
                ans.append(cur_level[::-1])
        return ans